### PR TITLE
arrow-cpp: 0.15.1 -> 0.16.0

### PR DESCRIPTION
**Pull request by tobim**
Changelog: https://arrow.apache.org/release/0.16.0.html

This switches to explicitly enabling dependency lookups for optional features (they were not picked up before).
It also enables building of the Parquet command line tools.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veprbl** [comment]
@GrahamcOfBorg build arrow-cpp
@GrahamcOfBorg build pkgsStaic.arrow-cpp
@GrahamcOfBorg build python37Packages.arrow-cpp
@GrahamcOfBorg build python38Packages.arrow-cpp

**Comment by veprbl** [comment]
@GrahamcOfBorg build pkgsStatic.arrow-cpp

**Comment by veprbl** [comment]
@GrahamcOfBorg build python37Packages.pyarrow
@GrahamcOfBorg build python38Packages.pyarrow

**Comment by veprbl** [comment]
Looks like this breaks pyarrow and the jemalloc-disable-shared.patch needs to be updated too.

**Comment by veprbl** [comment]
@GrahamcOfBorg build pkgsStatic.arrow-cpp
@GrahamcOfBorg build python37Packages.pyarrow
@GrahamcOfBorg build python38Packages.pyarrow

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 1]
```diff
@@ -35,6 +35,7 @@ in stdenv.mkDerivation rec {
   patches = [
     # patch to fix python-test
     ./darwin.patch
+    ./use-upstream-target-names.patch
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix]: This has to do with the enabling of CMAKE_FIND_PACKAGE_PREFER_CONFIG? We need CMAKE_FIND_PACKAGE_PREFER_CONFIG for the static build? Can we upstream this patch?

**Review by jonringer — APPROVED** [review]
diff LGTM

failures are broken on master
```
[15 built (4 failed), 0.0 MiB DL]
error: build of '/nix/store/fqr9xrpbj4qysrkml9v8b72haa4fc759-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/79817
4 package failed to build:
python37Packages.google_cloud_bigquery python37Packages.ibis-framework python38Packages.google_cloud_bigquery python38Packages.ibis-framework

16 package built:
arrow-cpp python27Packages.awkward python27Packages.pyarrow python27Packages.pytest-lazy-fixture python27Packages.uproot python27Packages.uproot-methods python37Packages.awkward python37Packages.pyarrow python37Packages.pytest-lazy-fixture python37Packages.uproot python37Packages.uproot-methods python38Packages.awkward python38Packages.pyarrow python38Packages.pytest-lazy-fixture python38Packages.uproot python38Packages.uproot-methods
```

File: pkgs/development/libraries/arrow-cpp/default.nix (hunk 1 above)
Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix]: I was planning to, just didn't get around to it yet.

Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix]: Upstreaming at https://github.com/apache/arrow/pull/6480.

**Comment by jonringer** [comment]
@GrahamcOfBorg build arrow-cpp python27Packages.awkward python27Packages.pyarrow python27Packages.pytest-lazy-fixture python27Packages.uproot python27Packages.uproot-methods python37Packages.awkward python37Packages.pyarrow python37Packages.pytest-lazy-fixture python37Packages.uproot python37Packages.uproot-methods python38Packages.awkward python38Packages.pyarrow python38Packages.pytest-lazy-fixture python38Packages.uproot python38Packages.uproot-methods

File: pkgs/development/libraries/arrow-cpp/default.nix (hunk 1 above)
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix]:
```suggestion
    (fetchpatch {
      url = "https://github.com/apache/arrow/commit/396861b38d2f4e805db7c2ecd2c96fff0ca2678b.patch";
      sha256 = "19hdm9ampq8g2swnyfgq7im4pqa55yy4ni86fjrb9yv5jzwvx0b1";
    })
```

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 2]
```diff
@@ -40,6 +40,15 @@ in stdenv.mkDerivation rec {
     ./jemalloc-disable-shared.patch
   ];
 
+  # Adjust CMake target names to make -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON work.
+  # Remove this when updating to the next version.
+  postPatch = ''
+    patch -p2 < ${(fetchpatch {
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix]: Just use
```
    (fetchpatch {
      url = "https://github.com/apache/arrow/commit/396861b38d2f4e805db7c2ecd2c96fff0ca2678b.patch";
      sha256 = "0ki7nx858374anvwyi4szz5hgnnzv4fghdd05c38bzry9rfljgb1";
      stripLen = 1;
    })
```

Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix]: Nice, I didn't know about that.

**Comment by veprbl** [comment]
@GrahamcOfBorg build pkgsStatic.arrow-cpp
@GrahamcOfBorg build python37Packages.pyarrow
@GrahamcOfBorg build python38Packages.pyarrow

**Review by veprbl — APPROVED** [review]

**Comment by veprbl** [comment]
@tobim Thank you!

**Pull request merged**